### PR TITLE
Add progress bar as a callback

### PR DESCRIPTION
**Pull request by nicofirst1**
Regarding [this issue](https://github.com/facebookresearch/EGG/issues/175)

### CODE REVIEW

**Comment by nicofirst1** [comment]
The tests fail because some error in how the len is defined in [simple_autoenc](https://github.com/facebookresearch/EGG/tree/d9d027d0595b36ff02ff124d22c44a205d728457/egg/zoo/simple_autoenc), [channel](https://github.com/facebookresearch/EGG/tree/d9d027d0595b36ff02ff124d22c44a205d728457/egg/zoo/channel) and [lanaguae bottleneck](https://github.com/facebookresearch/EGG/tree/d9d027d0595b36ff02ff124d22c44a205d728457/egg/zoo/language_bottleneck).

I believe they all implemented the same [onehot loader](https://github.com/facebookresearch/EGG/blob/d9d027d0595b36ff02ff124d22c44a205d728457/egg/zoo/language_bottleneck/guess_number/features.py#L76) which does not have the `_dataset_kind` attribute. 

Rather than catch this exception I would correct those game.

**Comment by eugene-kharitonov** [comment]
Oh, there is a lot here! :) 
Could you please unpack in different PRs those two changes? It would be faster to proceed, cause /1/ is trivial and /2/ might deserve some discussion:
/1/ Adding on_batch
/2/ Adding tqdm part (preferably in a separate logger; also you'd need to add tqdm to requirements.txt?)
Also
3/ There are lots of indentation changes, maybe they are valid, but it would be better not to mix changes to codestyle/formatting in random places with meaningful changes.
4/ Also there are pieces of another PR, I think? Could you please rebase this branch on top of master? Otherwise it could be messy-ish.

File: egg/core/callbacks.py [hunk 1]
```diff
@@ -69,6 +79,24 @@ def on_test_end(self, loss: float, logs: Interaction, epoch: int):
     def on_epoch_end(self, loss: float, logs: Interaction, epoch: int):
         if self.print_train_loss:
             self.aggregate_print(loss, logs, "train", epoch)
+        self.batch_info['cur_epoch'] += 1
+        self.batch_info['cur_batch'] = 0
+        print()
```
Comment (robertodessi) [inline egg/core/callbacks.py:84]: ?

File: egg/core/callbacks.py [hunk 2]
```diff
@@ -44,11 +43,22 @@ def on_epoch_begin(self, epoch: int):
...
+    def on_batch(self, logs: Interaction, loss: float, data_len: int):
+        pass
+
 
 class ConsoleLogger(Callback):
-    def __init__(self, print_train_loss=False, as_json=False):
+    def __init__(self, print_train_loss=False, as_json=False, print_batch_progress=False, n_epochs=-1):
         self.print_train_loss = print_train_loss
         self.as_json = as_json
+        self.print_batch_progress = print_batch_progress
```
Comment (robertodessi) [inline egg/core/callbacks.py:54]: This is never used

File: egg/core/callbacks.py [hunk 3]
```diff
@@ -69,6 +79,24 @@ def on_test_end(self, loss: float, logs: Interaction, epoch: int):
     def on_epoch_end(self, loss: float, logs: Interaction, epoch: int):
         if self.print_train_loss:
             self.aggregate_print(loss, logs, "train", epoch)
+        self.batch_info['cur_epoch'] += 1
+        self.batch_info['cur_batch'] = 0
+        print()
+
+    def on_batch(self, logs: Interaction, loss: float, data_len: int):
+
+        bi = self.batch_info
```
Comment (robertodessi) [inline egg/core/callbacks.py:88]: reason to have bi saved here? Plus I think we would a better name for the variable

File: egg/core/callbacks.py [hunk 4]
```diff
@@ -69,6 +79,24 @@ def on_test_end(self, loss: float, logs: Interaction, epoch: int):
...
+        elapsed = time.time() - bi['batch_time']
+
+        to_print = f"\rEpoch {bi['cur_epoch']}"
+        if bi['n_epochs'] > 0: to_print += f"/{bi['n_epochs']}"
```
Comment (robertodessi) [inline egg/core/callbacks.py:92]: Please check with flake8 as well https://www.flake8rules.com/rules/E701.html

File: egg/core/callbacks.py [hunk 5]
```diff
@@ -69,6 +79,24 @@ def on_test_end(self, loss: float, logs: Interaction, epoch: int):
...
+        to_print += "| train: "
+        to_print += self.tqdm.format_meter(n=bi['cur_batch'], total=data_len, elapsed=elapsed)
```
Comment (robertodessi) [inline egg/core/callbacks.py:94]: single quotes are not admitted in black-style

File: egg/core/callbacks.py [hunk 6]
```diff
@@ -69,6 +79,24 @@ def on_test_end(self, loss: float, logs: Interaction, epoch: int):
...
+        print(to_print, end="")
```
Comment (robertodessi) [inline egg/core/callbacks.py:95]: This cannot be disabled which is not ideal for cluster-based runs

File: egg/core/callbacks.py [hunk 7]
```diff
@@ -44,11 +43,22 @@ def on_epoch_begin(self, epoch: int):
     def on_epoch_end(self, loss: float, logs: Interaction, epoch: int):
         pass
 
+    def on_batch(self, logs: Interaction, loss: float, data_len: int):
+        pass
+
 
 class ConsoleLogger(Callback):
-    def __init__(self, print_train_loss=False, as_json=False):
+    def __init__(self, print_train_loss=False, as_json=False, print_batch_progress=False, n_epochs=-1):
```
Comment (robertodessi) [inline egg/core/callbacks.py:51]: Reasons to start from -1? And I'm thinking ConsoleLogger and ProgressBarUpdater should be two separate callbacks

File: egg/core/callbacks.py [hunk 8]
```diff
@@ -2,16 +2,15 @@
 
 # This source code is licensed under the MIT license found in the
 # LICENSE file in the root directory of this source tree.
-
```
Comment (robertodessi) [inline egg/core/callbacks.py:5]: ?

File: egg/core/trainers.py [hunk 9]
```diff
@@ -193,12 +199,14 @@ def train_epoch(self):
...
-                self.distributed_context.is_distributed
-                and self.aggregate_interaction_logs
+                    self.distributed_context.is_distributed
+                    and self.aggregate_interaction_logs
             ):
                 interaction = Interaction.gather_distributed_interactions(interaction)
             interaction = interaction.to("cpu")
             interactions.append(interaction)
+            for callback in self.callbacks:
+                callback.on_batch(interaction, optimized_loss.detach(), data_len)
```
Comment (robertodessi) [inline egg/core/trainers.py:209]: I don't think detach is necessary here

**Comment by robertodessi** [comment]
CI tests are failing since tqdm is not being imported. If it's added to the requirements file it should be fine

**Comment by nicofirst1** [comment]
I see now that I was not on the master branch, I'm creating two other PRs with cleaner version (@eugene-kharitonov ).
I'm not closing this yet so we can still use @robertodessi 's work
